adapters/akka-http/src/main/scala/caliban/AkkaHttpAdapter.scala

```diff
@@ -113,24 +117,45 @@ trait AkkaHttpAdapter {
                   .forkDaemon
                   .flatMap(fiber => subscriptions.update(_.updated(message.id, fiber)))
               case other =>
-                sendMessage(sendTo, message.id, other, result.errors) *> IO.fromFuture(_ =>
-                  sendTo.offer(TextMessage(s"""{"type":"complete","id":"${message.id}"}"""))
-                )
+                sendMessage(sendTo, message.id, other, result.errors) *> IO
+                  .fromFuture(_ => sendTo.offer(TextMessage(s"""{"type":"complete","id":"${message.id}"}""")))
             }
       } yield ()
 
     get {
       extractUpgradeToWebSocket { upgrade =>
-        val (queue, source) = Source.queue[Message](0, OverflowStrategy.fail).preMaterialize()
-        val subscriptions   = runtime.unsafeRun(Ref.make(Map.empty[String, Fiber[Throwable, Unit]]))
+        val (queue: SourceQueueWithComplete[Message], source) =
+          Source.queue[Message](0, OverflowStrategy.fail).preMaterialize()
+        val subscriptions = runtime.unsafeRun(Ref.make(Map.empty[String, Fiber[Throwable, Unit]]))
         val sink = Sink.foreach[Message] {
           case TextMessage.Strict(text) =>
             val io = for {
               msg     <- Task.fromEither(json.parseWSMessage(text))
               msgType = msg.messageType
               _ <- IO.whenCase(msgType) {
                     case "connection_init" =>
-                      IO.fromFuture(_ => queue.offer(TextMessage("""{"type":"connection_ack"}""")))
+                      val ack: ZIO[Clock with Random, Throwable, Any] =
+                        IO.fromFuture(_ => queue.offer(TextMessage("""{"type":"connection_ack"}""")))
+
+                      //If we wanted a keepAlive message to be sent, we set it up here,
+                      //note that we save it with a special id in the subscriptions as yet another
+                      //fiber that we have to interrupt later.
+                      val withKeepAlive = keepAliveTime.fold(ack) { time =>
+                        ack.flatMap(_ =>
+                          IO.fromFuture(_ =>
+                              queue
+                                .offer(
+                                  TextMessage("""{"type":"ka"}""")
+                                )
+                            )
+                            .repeat(Schedule.spaced(time).jittered)
+                            .forkDaemon
+                            .map(keepAliveFiber =>
+                              subscriptions.update(_.updated("_keepAliveFiber", keepAliveFiber.unit))
```

Comment:
Maybe a map of `Option[String]` would be more elegant? And we use `None` for this case.

Reply:
That works.